### ISSUE TITLE
Generate for dependencies of specific classes

### ISSUE THREAD

**Issue by michaelaird** [frhagn/Typewriter#213]
I've been bashing my head against this for most of the day and I can't figure it out. I'm hoping it's even possible.

Here's the scenario: I want to generate a service for specific WebApi controllers and also generate interfaces for the supporting DTOs and enums that are used by those controllers. I don't want to  generate interfaces for all DTOs and enums, just the ones that are actually used by the controllers i've specified.

I understand the syntax for filtering $Classes on the names of the controllers I care about but I can't figure out how to parse through those specific classes to get the dependencies (method parameters and return types) to generate the interfaces.

Any help or suggestions would be greatly appreciated. (This project is amazing!)

**Comment by mahmoudelgawad** [comment]
May the following link help you:
https://stackoverflow.com/questions/41552602/typewriter-render-from-list-of-c-sharp-class-names

**Comment by michaelaird** [comment]
Yep. I’m using a lambda to specify the API controller names. I’m then
trying to inspect the methods in those controllers and render interfaces
for only the dtos used by the controllers.

I suppose I could hardcode the list but I’m trying to do it dynamically.

**Comment by Marcel1024** [comment]
this should get you started:

```

 string GetUsedTypes(Class c){
        HashSet<string> usedTypes = new HashSet<string>();
        foreach(Method m in c.Methods)
        {
            var returnType = m.Type;
            if(returnType.IsEnumerable){
            returnType = returnType.TypeArguments[0];
            }
            
            if(!returnType.IsPrimitive && 
               !returnType.Name.Equals("void", StringComparison.OrdinalIgnoreCase) &&
               !returnType.Name.Equals("IActionResult", StringComparison.OrdinalIgnoreCase)
               )
            {
                usedTypes.Add(returnType);
            }

            foreach(var parameter in m.Parameters)
            {
                if(!parameter.Type.IsPrimitive){
                    usedTypes.Add(parameter.Type.Name);
                }
            }
        }

        string result = string.Empty;
        foreach(string usedType in usedTypes)
        {
            if(!String.IsNullOrEmpty(result))
            {
                result += Environment.NewLine;
            }

            result += $"import {usedType} from './models/{usedType}.d.ts';";
        }

        return result;
    }

```

Then use:
```

$Classes(:Controller)[
$GetUsedTypes
]
```

**Comment by Marcel1024** [comment]
oops, posted this to the wrong question.

**Comment by nnic** [comment]
I've also been trying to do this, and and I have found a way, however, it will only create one file with all the interfaces and enums in, and will only look one level deep, but this can be extended. 

`${
    using Typewriter.Extensions.Types;

    bool ClassFilter(Class c) => c.FullName == "Namespace.ViewModel";
}
$Classes(c => ClassFilter(c))[
export interface $Name {$Properties[
    $Name: $Type;]
}
$Properties(x => x.Type.IsDefined)[$Type[
export interface $Name {$Properties[
    $Name: $Type;]
}
$Properties(x => x.Type.IsDefined)[$Type[
export interface $Name {$Properties[
    $Name: $Type;]
}]]
$Properties(x => x.Type.IsGeneric && !x.Type.IsPrimitive)[$Type[
export interface $Unwrap[$Name {$Properties[
    $Name: $Type;]]
}]]
$Properties(x => x.Type.IsEnum)[$Type[
export enum $Name {$Constants[
    $Name = $Value][,]
}]]]]
$Properties(x => x.Type.IsGeneric && !x.Type.IsPrimitive)[$Type[
export interface $Unwrap[$Name {$Properties[
    $Name: $Type;]]
}]]
$Properties(x => x.Type.IsEnum)[$Type[
export enum $Name {$Constants[
    $Name = $Value][,]
}
]]
]`

Basically the code gets the initial class and creates an interface with all the properties. Next it looks at the type of each property, and if the type has been defined in the solution, then it will create an interface based on that type. Next it will do the same for any generic types, and finally for enums. The code then does the same to each of the property types once again, so only goes one level deep. You can extend this by copying code to add a greater depth, but I'm not sure there is a way to do it recursively.

**Comment by michaelaird** [comment]
Interesting. That might be a viable solution. If the same dependency gets used multiple times it might show up multiple times but TypeScript will just combine them. Some kind of recursion on the template would be useful....

**Comment by michaelaird** [comment]
@nnic your approach gave me some ideas and I took another crack at it tonight. I have something 95% working but I'm struggling with bringing in base types (I've opened issue 221 for that).  Here's a gist of what I have so far : https://gist.github.com/michaelaird/cfd15cebe92585556be2a14abff9dff4

**Comment by michaelaird** [comment]
With the latest release, my issues are solved.

**Comment by alexmnv** [comment]
> With the latest release, my issues are solved.

Could you please share your solution with us?
